src/udev/iocost/iocost.c

```diff
@@ -0,0 +1,314 @@
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        r = query_named_solution(device, name, &model_params, &qos_params);
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_devnum(device, &devnum);
+        if (r < 0)
+                return log_error_errno(r, "Error getting devnum for device %s: %m", path);
+
+        if (asprintf(&qos, DEVNUM_FORMAT_STR " enable=1 ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), qos_params) < 0)
+                return log_oom();
+
+        if (asprintf(&model, DEVNUM_FORMAT_STR " model=linear ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), model_params) < 0)
+                return log_oom();
+
+        log_debug("Applying iocost parameters to %s using solution '%s'\n"
+                        "\tio.cost.qos: %s\n"
+                        "\tio.cost.model: %s\n", path, name ?: "default", qos, model);
+
+        r = cg_set_attribute("io", NULL, "io.cost.qos", qos);
+        if (r < 0) {
+                log_full_errno(r == -ENOENT ? LOG_DEBUG : LOG_ERR, r, "Failed to set io.cost.qos: %m");
+                return r == -ENOENT ? 0 : r;
+        }
+
+        r = cg_set_attribute("io", NULL, "io.cost.model", model);
+        if (r < 0) {
+                log_full_errno(r == -ENOENT ? LOG_DEBUG : LOG_ERR, r, "Failed to set io.cost.model: %m");
+                return r == -ENOENT ? 0 : r;
+        }
+
+        return 0;
+}
+
+static int query_solutions_for_path(const char *path) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_strv_free_ char **solutions = NULL;
+        const char *default_solution = NULL;
+        const char *model_name = NULL;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        if ((sd_device_get_property_value(device, "ID_MODEL_FROM_DATABASE", &model_name) < 0 &&
+             sd_device_get_property_value(device, "ID_MODEL", &model_name) < 0))
+                return log_error_errno(SYNTHETIC_ERRNO(ENOENT), "Model name for device %s is unknown", path);
+
+        r = get_known_solutions(device, &solutions);
+        if (r < 0)
+                return log_error_errno(r, "No solutions found for device %s, model name %s on hwdb: %m\n", path, model_name);
+
+        r = choose_solution(solutions, &default_solution);
+        if (r < 0)
+                return log_error_errno(
+                                r,
+                                "No solutions found for device %s, model name %s on hwdb: %m",
+                                path,
+                                model_name);
+
+        log_info("Known solutions for %s model name: %s\n", path, model_name);
+        log_info("Preferred bandwidth threshold: %s", arg_bandwidth_threshold);
+        log_info("Solution that would be applied: %s", default_solution);
```

Comment:
please use a multi-line format str, and issue this in one message.

i.e.:

```c
log_info("Known … name: %s\n"
         "Preferred … threshold: %s\n"
         "Solution … applied: %s\n",
         paht, model_name, 
         arg_bandwidth_threshold,
         …
```

you get the idea.

this will then end up as one entry in the journal, and given this belongs together that's a good idea.